### ISSUE TITLE
SquashFS regression

### ISSUE THREAD

**Issue by oschulz** [apptainer/singularity#817]
WIth singularity v2.3.1-dist, I have trouble running some (not all) SquashFS images. The images run fine with v2.2.1, but with v2.3.1 I get

    ERROR  : Failed to mount squashfs image in (read only): Invalid argument
    ABORT  : Retval = 255

I can provide a container for testing - it's quite big, though (3.7 GB SquashFS image).

**Comment by oschulz** [comment]
```singularity -vvv [...]``` says

```
VERBOSE: File is not a valid Singularity image
VERBOSE: Calculating image offset
VERBOSE: Found image at an offset of 9 bytes
VERBOSE: Found available loop device: /dev/loop0
VERBOSE: Using loop device: /dev/loop0
VERBOSE: Set container directory to: /var/lib/singularity/mnt/container
VERBOSE: Checking what kind of image we are mounting
VERBOSE: Attempting to mount as squashfs image
VERBOSE: Mounting squashfs image
ERROR  : Failed to mount squashfs image in (read only): Invalid argument
ABORT  : Retval = 255
```

Since c78bc4f, calculating offsets for SquashFS images should be prevented in "offset.c" by

```
    if ( singularity_image_check(image) != 0 ) {
        singularity_message(DEBUG, "File is not a Singularity image, returning zero offset\n");
        return(0);
    }
```

but somehow it doesn't seem to work anymore. Did the behaviour of `singularity_image_check(image)` change?

**Comment by oschulz** [comment]
This is a serious problem for us, currently (due to our recent upgrade to v2.3.1). We'll have to downgrade to v2.2.1, temporarily, until this is fixed. :-(

**Comment by vsoch** [comment]
hey @oschulz ! I'm not super familiar with squashfs, but could you provide the entire debug output (as a file would be good since it's likely super long) and I can trace the evocation pathway and guess what might be up?

**Comment by oschulz** [comment]
Thanks, @vsoch  - it's not that long, actually:

```
$ singularity -vvv shell myimage.sqsh 
Increasing verbosity level (4)
Singularity version: 2.3.1-dist
Exec'ing: /usr/lib/x86_64-linux-gnu/singularity/cli/shell.exec
Evaluating args: 'myimage.sqsh'
VERBOSE: Set messagelevel to: 4
VERBOSE: Initialize configuration file: /etc/singularity/singularity.conf
VERBOSE: Got config key allow setuid = 'yes'
VERBOSE: Got config key max loop devices = '256'
VERBOSE: Got config key allow pid ns = 'yes'
VERBOSE: Got config key config passwd = 'yes'
VERBOSE: Got config key config group = 'yes'
VERBOSE: Got config key config resolv_conf = 'yes'
VERBOSE: Got config key mount proc = 'yes'
VERBOSE: Got config key mount sys = 'yes'
VERBOSE: Got config key mount dev = 'yes'
VERBOSE: Got config key mount home = 'yes'
VERBOSE: Got config key mount tmp = 'yes'
VERBOSE: Got config key mount hostfs = 'no'
VERBOSE: Got config key bind path = '/etc/localtime'
VERBOSE: Got config key bind path = '/etc/hosts'
VERBOSE: Got config key user bind control = 'yes'
VERBOSE: Got config key enable overlay = 'yes'
VERBOSE: Got config key mount slave = 'yes'
VERBOSE: Got config key sessiondir max size = '16'
VERBOSE: Initializing Singularity Registry
VERBOSE: Adding value to registry: 'LIBEXECDIR' = '/usr/lib/x86_64-linux-gnu'
VERBOSE: Adding value to registry: 'COMMAND' = 'shell'
VERBOSE: Adding value to registry: 'MESSAGELEVEL' = '4'
VERBOSE: Adding value to registry: 'VERSION' = '2.3.1-dist'
VERBOSE: Adding value to registry: 'LOCALSTATEDIR' = '/var/lib'
VERBOSE: Adding value to registry: 'SYSCONFDIR' = '/etc'
VERBOSE: Adding value to registry: 'BINDIR' = '/usr/bin'
VERBOSE: Adding value to registry: 'IMAGE' = 'myimage.sqsh'
VERBOSE: Set home (via getpwuid()) to: /home/[...]
VERBOSE: Running SUID program workflow
VERBOSE: Checking program has appropriate permissions
VERBOSE: Checking configuration file is properly owned by root
VERBOSE: Checking if singularity.conf allows us to run as suid
VERBOSE: Invoking the user namespace
VERBOSE: Not virtualizing USER namespace: running as SUID
VERBOSE: Not virtualizing PID namespace on user request
VERBOSE: Not virtualizing IPC namespace on user request
VERBOSE: Using session directory: /var/lib/singularity/mnt/session
VERBOSE: Adding value to registry: 'SESSIONDIR' = '/var/lib/singularity/mnt/session'
VERBOSE: Checking what kind of image we are mounting
VERBOSE: Checking that file pointer is a Singularity image
VERBOSE: File is not a valid Singularity image
VERBOSE: Checking what kind of image we are mounting
VERBOSE: Checking that file pointer is a Singularity image
VERBOSE: File is not a valid Singularity image
VERBOSE: Calculating image offset
VERBOSE: Found image at an offset of 9 bytes
VERBOSE: Found available loop device: /dev/loop0
VERBOSE: Using loop device: /dev/loop0
VERBOSE: Set container directory to: /var/lib/singularity/mnt/container
VERBOSE: Checking what kind of image we are mounting
VERBOSE: Attempting to mount as squashfs image
VERBOSE: Mounting squashfs image
ERROR  : Failed to mount squashfs image in (read only): Invalid argument
ABORT  : Retval = 255
```

**Comment by vsoch** [comment]
could you add `--debug` as well?

**Comment by oschulz** [comment]
Oh, sorry, sure:

[singularity-log.txt](https://github.com/singularityware/singularity/files/1160441/singularity-log.txt)

The problem seems to be that Singularity thinks it's detected a Singularity image header, and therefore calculates an image offset for mounting. As SquashFS files don't have Singularity headers, the mount fails, obviously. This is why we disabled offsets for SquashFS and similar in c78bc4f, only it doesn't seem to work anymore. Singularity does detect that it's not a Singularity image and reports

    File is not a valid Singularity image

but `singularity_image_check(image) != 0` doesn't seem to get triggered, so instead of the expected

    File is not a Singularity image, returning zero offset

we get

    Calculating image offset

**Comment by oschulz** [comment]
I'm not familiar with the changes since v2.2.1, but from what I can see, the semantics of `singularity_image_check` have changed:

In v2.2.1, it was `singularity_image_check(FILE *image_fp)` and returned `-1` if the file was not a valid Singularity image.

Now it's `singularity_image_check(struct image_object *image)`, implemented by `_singularity_image_check(struct image_object *image)`, and always returns `0` for known image formats, but sets `image->type`. So maybe instead of

    if ( singularity_image_check(image_fp) != 0 ) {
        singularity_message(DEBUG, "File is not a Singularity image, returning zero offset\n");
        return(0);
    }

(used to be in `image-util.c`, was moved to  "offset.c") it should be

    if ( image->type == SQUASHFS ) {
        singularity_message(DEBUG, "File is a SquashFS image, returning zero offset\n");
        return(0);
    }

now?

**Comment by oschulz** [comment]
Here's the proposed fix: #820

**Comment by vsoch** [comment]
Thanks :) 

Ok, so based on what you are saying, what is supposed to happen is that squashFS images are special types that should return a 0 offset.  I think this might be a bug with regard to the fact that we are mounting an image, because it's technically creating a "session directory" for it. I have no clue if this is helpful, but here is what I was able to dig into. The first 

Specifically, when this happens if a default size isn't set, it defaults to what is in the config:

```
DEBUG   [U=1001,P=6349]    singularity_config_get_value_impl()       Returning configuration value sessiondir max size='16'
```

and then [-7 is calculated from that](https://github.com/singularityware/singularity/blob/adf5259ac94d7b213c649f073dcda32459b0c8b7/src/util/sessiondir.c#L78):

```
DEBUG   [U=1001,P=6349]    singularity_sessiondir()                  Creating the sessiondir size mount option length
DEBUG   [U=1001,P=6349]    singularity_sessiondir()                  Got size length of: 9
```

Then we get the image:

```
 image = singularity_image_init(singularity_registry_get("IMAGE"));
DEBUG   [U=1001,P=6349]    singularity_registry_get()                Returning value from registry: 'IMAGE' = 'myimage.sqsh'
```

and is it squashfs? This is where weirdness starts - it goes through this sequence **twice** but only identifies squashfs the second time.

```
DEBUG   [U=1001,P=6349]    singularity_image_open()                  Opening file descriptor to image: myimage.sqsh
```
the first time, the call goes like singularity_image_check() --> _singularity_image_check() and then there is a function to check for [different types](https://github.com/singularityware/singularity/blob/master/src/lib/image/check/check.c#L51) and specifically here we should be returning 0 and seeing that it is squashfs:

```
    if ( _singularity_image_check_image(image) == 0 ) {
        image->type = SINGULARITY;
    } else if ( _singularity_image_check_squashfs(image) == 0 ) {
        image->type = SQUASHFS;
    } else if ( _singularity_image_check_dir(image) == 0 ) {
        image->type = DIRECTORY;
    } else {
        singularity_message(ERROR, "Unknown image format/type.\n");
        ABORT(255);
    }
```
first we see it's not the first (SINGULARITY)

```
VERBOSE [U=1001,P=6349]    singularity_image_check()                 Checking what kind of image we are mounting
VERBOSE [U=1001,P=6349]    singularity_image_check_image()           Checking that file pointer is a Singularity image
DEBUG   [U=1001,P=6349]    singularity_image_check_image()           First line of image(fd=4): hsqs�w
```
Here we are looking at the LAUNCH_STRING, which is the first line of a singularity image, and comparing to what the image has:
```
DEBUG   [U=1001,P=6349]    singularity_image_check_image()           Checking if first line matches key
VERBOSE [U=1001,P=6349]    singularity_image_check_image()           File is not a valid Singularity image
```
meaning we are comparing these two things (not the same):

```
LAUNCH_STRING "#!/usr/bin/env run-singularity\n"
DEBUG   [U=1001,P=6349]    singularity_image_check_image()           uFirst line of image(fd=4): hsqs�w
```
so that function returns -1, and then we would expect the next call to check for squashfs, [here](https://github.com/singularityware/singularity/blob/master/src/lib/image/check/squashfs/squashfs.c):

And the function is only verbose given that it's NOT squashfs:

```
int _singularity_image_check_squashfs(struct image_object *image) {
    char *image_name = strdup(image->name);
    int len = strlength(image_name, 1024);

    if ( strcmp(&image_name[len-5], ".sqsh") != 0 ) {
        singularity_message(DEBUG, "Image does not appear to be of type '.sqsh': %s\n", image->path);
        return(-1);
    }

    return(0);
}
```
 so based on this and your output, I would expect it was detected, and the variable for image type set to SQUASHFS ([which is 2](https://github.com/singularityware/singularity/blob/c12d9844fe98b4609171acc3a16066f1a4abff9c/src/lib/image/image.h#L28)) However, I don't see this reflected in the debug output, it jumps right to getting a value from the config and then trying to mount:

```
DEBUG   [U=1001,P=6349]    singularity_config_get_value_impl()       Returning configuration value max loop devices='256'
DEBUG   [U=1001,P=6349]    singularity_image_bind()                  Entered singularity_image_bind()
```
but now it gets confusing, because it sets everything to read only, and goes through the exact same sequence again:

```
DEBUG   [U=1001,P=6349]    singularity_image_bind()                  Setting loopdev open to: O_RDONLY
```

but then it fails because it's read only.

```
VERBOSE [U=1001,P=6349]    singularity_image_mount()                 Checking what kind of image we are mounting
VERBOSE [U=1001,P=6349]    singularity_image_mount()                 Attempting to mount as squashfs image
DEBUG   [U=1001,P=6349]    singularity_priv_escalate()               Temporarily escalating privileges (U=1001)
DEBUG   [U=0,P=6349]       singularity_priv_escalate()               Clearing supplementary GIDs.
VERBOSE [U=0,P=6349]       singularity_image_mount_squashfs_mount()  Mounting squashfs image
ERROR   [U=0,P=6349]       singularity_image_mount_squashfs_mount()  Failed to mount squashfs image in (read only): Invalid argument
ABORT   [U=0,P=6349]       singularity_image_mount_squashfs_mount()  Retval = 255
```
So I (think?) what might be happening is that the variable is set for SQUASHFS, but then it's not carried forward for the overlay check.  We could minimally put more verbose output messages in that section to know that the original detection was squash fs (right now it's detected silently) 

@gmkurtzer - why is this whole sequence done twice?

**Comment by oschulz** [comment]
> Ok, so based on what you are saying, what is supposed to happen is that squashFS images are special types that should return a 0 offset.

Well, I wouldn't say they're special, it's just that SquashFS doesn't have Singularity-style image labels. But depending in the first bytes in the SquashFS file, Singularity's offset detection mechanism can be triggered, by random chance, and result in a bogus offset - so offset detection has to be disabled for SquashFS.

I don't know about the rest, but I haven't had any other troubles with SquashFS yet.

**Comment by oschulz** [comment]
Fixed by #820
